src/borg/_hashindex.c

```diff
@@ -369,49 +446,129 @@ hashindex_write(HashIndex *index, const char *path)
 static const void *
 hashindex_get(HashIndex *index, const void *key)
 {
-    int idx = hashindex_lookup(index, key);
+    int idx = hashindex_lookup(index, key, NULL);
     if(idx < 0) {
         return NULL;
     }
     return BUCKET_ADDR(index, idx) + index->key_size;
 }
 
+inline void
+/* memswap(uint8_t a[], uint8_t b[], void *tmp, size_t entry_size) { */
+/*     uint8_t t; */
+/*     unsigned int i; */
+/*     for (i=0; i<entry_size; i++){ */
+/* 	t = a[i]; */
+/* 	a[i] = b[i]; */
+/* 	b[i] = t; */
+/*     } */
+memswap(void *a, void *b, void *tmp, size_t entry_size) {
+    memcpy(tmp, a, entry_size);
+    memcpy(a, b, entry_size);
+    memcpy(b, tmp, entry_size);
+}
+
 static int
 hashindex_set(HashIndex *index, const void *key, const void *value)
 {
-    int idx = hashindex_lookup(index, key);
-    uint8_t *ptr;
-    if(idx < 0)
+    int offset = 0;
+    int idx = hashindex_lookup(index, key, &offset);
+    uint8_t *bucket_ptr;
+    int other_offset;
+    int entry_size = (index->key_size + index->value_size);
+    /* uint8_t entry_to_insert[entry_size]; */
+    void *entry_to_insert = index->entry_to_insert;
+    /* uint8_t tmp_entry[entry_size]; */
+    void *tmp_entry = index->tmp_entry;
+    if(idx >= 0)
+    {
+        /* we already have the key in the index we just need to update its value */
+        memcpy(BUCKET_ADDR(index, idx) + index->key_size, value, index->value_size);
+        if (DEBUG) {
+            updates++;
+        }
+    }
+    else
     {
+        /* we don't have the key in the index we need to find an appropriate address */
+        if (DEBUG) {
+            inserts ++;
+        }
         if(index->num_entries > index->upper_limit) {
+            /* we need to grow the hashindex */
             if(!hashindex_resize(index, grow_size(index->num_buckets))) {
                 return 0;
             }
+            offset = 0;
+        }
+        idx = hashindex_index(index, key) + offset;
+        if (idx >= index->num_buckets){
+            idx = idx - index->num_buckets;
+        }
+        memcpy(entry_to_insert, key, index->key_size);
+        memcpy(entry_to_insert + index->key_size, value, index->value_size);
+        bucket_ptr = BUCKET_ADDR(index, idx);
+        if (DEBUG) {
+            collisions += offset;
         }
-        idx = hashindex_index(index, key);
         while(!BUCKET_IS_EMPTY(index, idx) && !BUCKET_IS_DELETED(index, idx)) {
-            idx = (idx + 1) % index->num_buckets;
+            /* we have a collision */
```

Comment:
So basically take the block (the entire contiguous section of buckets up to the first empty bucket) and `memmove` it forward one address, then insert at the ideal location. Oh my, yes, such an elegant solution! Thanks!

> not on their absolute value (other_offset < offset).

Not sure what you mean by absolute value. The offsets are the relative distance from the ideal bucket a key would be in vs which bucket it is now, so I think it's the same as what you call "displacement".

> Tell me if something in my explanation is unclear. I'm using different names. My word for "distance" is "displacement".

Indeed, displacement is a better name for it.


Thanks for the feedback!